core/engine.go

```diff
@@ -195,27 +182,31 @@ func (e *Engine) startBackgroundProcesses(
 		case err = <-execRunResult:
 			if err != nil {
 				e.logger.WithError(err).Debug("run: execution scheduler returned an error")
-				e.setRunStatusFromError(err)
 			} else {
 				e.logger.Debug("run: execution scheduler finished nominally")
-				e.OutputManager.SetRunStatus(lib.RunStatusFinished)
 			}
 			// do nothing, return the same err value we got from the Run()
 			// ExecutionScheduler result, we just set the run_status based on it
 		case <-runCtx.Done():
 			e.logger.Debug("run: context expired; exiting...")
-			e.OutputManager.SetRunStatus(lib.RunStatusAbortedUser)
-			err = errext.WithExitCodeIfNone(errors.New("test run aborted by signal"), exitcodes.ExternalAbort)
+			err = errext.WithAbortReasonIfNone(
+				errext.WithExitCodeIfNone(errors.New("test run aborted by signal"), exitcodes.ExternalAbort),
+				errext.AbortedByUser,
+			)
```

Comment:
This wrapping is a bit convoluted. It seems we're only wrapping exit code errors in aborted errors because it's convenient to return a single error here. But conceptually, these errors aren't related in this way. If anything, maybe there should be a mapping from `AbortReason` to `ExitCode`, which can happen outside of `core`. After all, why should the `Engine` be concerned with process exit codes?

I'm not sure I have a concrete improvement suggestion here, but this seems fishy in several ways. :thinking:

Reply:
Great point! Now that this PR adds the new `AbortReason` type that can accurately represent the internal reason for why a k6 test run failed, we can use it for external mappings. The PR does the mapping from `AbortReason` to `cloudapi.RunStatus`, but I don't see a reason why we couldn't also have a mapping to `ExitCode`. We'd probably need to add more `AbortReason` values, but that should be fine :thinking: In any case, I'd prefer to handle this in a separate PR, so I added this https://github.com/grafana/k6/issues/2804#issuecomment-1399978135